mobify/progressive-web-scaffold
# WEB-1029: Offline mode functionality

Adds a couple of components and some logic to allow for 'offline mode functionality' if a network request for a page fails (for whatever reason).

 **JIRA**: [WEB-1029](https://mobify.atlassian.net/browse/WEB-1029)

## Changes
- Add `Offline` and `OfflineBanner` stub components based on design
- `onPageReceived` reducer adds fetched URL to a new `state.app`-level keys `fetchedUrls` and `fetchError`
- `App` container uses `fetchedUrls` and `fetchError` keys to determine whether to display offline components
- Worker uses custom fetch handler for modifying response body of `//online.mobify.net/online.json` to `{offline:true}` when request fails

## Todo
- [x] Refactor after Data Architecture 2.0 is merged
- [x] Fix problem of Offline component not dispatching router/fetch actions properly
- [x] Styling of offline components
- ~Custom fetch handler cache expiry~
- Implement whitelist / blacklist ?

## How to test-drive this PR
- Preview the app
- Emulate offline (DevTools > Network > Check 'Offline' checkbox)
- Navigate back to home, see that offline banner displays
- Navigate to any other page, see that offline splash displays
